File: Sample/Koudai/Server/src/ZyGames.Tianjiexing.Server/PyScript/Action/action1023.py
import clr, sys
clr.AddReference('ZyGames.Framework.Common');
clr.AddReference('ZyGames.Framework');
clr.AddReference('ZyGames.Framework.Game');
clr.AddReference('ZyGames.Tianjiexing.Model');
clr.AddReference('ZyGames.Tianjiexing.BLL');
clr.AddReference('ZyGames.Tianjiexing.Lang');
clr.AddReference('ZyGames.Tianjiexing.Component');

from action import *
from System import *
from System.Collections.Generic import *
from lang import Lang
from ZyGames.Framework.Common.Log import *
from ZyGames.Tianjiexing.Model.ConfigModel import *
from ZyGames.Framework.Common import *
from ZyGames.Tianjiexing.Model import *
from ZyGames.Tianjiexing.BLL import *
from ZyGames.Tianjiexing.Lang import *
from ZyGames.Framework.Game.Cache import *
from ZyGames.Framework.Game.Service import *
from ZyGames.Framework.Game.Runtime import *
from ZyGames.Framework.Common import *
from ZyGames.Framework.Cache.Generic import *
from ZyGames.Tianjiexing.Model.Enum import *
from ZyGames.Tianjiexing.BLL.Base import *
from ZyGames.Tianjiexing.Component import *
from ZyGames.Framework.Common.Serialization import *
from ZyGames.Framework.Cache.Generic import *
from ZyGames.Framework.Game.Runtime import *
from ZyGames.Tianjiexing.Component.Chat import *
from ZyGames.Framework.Game.Com.Rank import *
from ZyGames.Tianjiexing.Model.Config import *
from ZyGames.Tianjiexing.BLL.Action import *

# 1023_玩家改名
class UrlParam(HttpParam):
    def __init__(self):
        HttpParam.__init__(self)
        self.nickName = ''

class ActionResult(DataResult):
    def __init__(self):
        DataResult.__init__(self)

def getUrlElement(httpGet, parent):
    urlParam = UrlParam();
    if httpGet.Contains("NickName"):
        urlParam.nickName = httpGet.GetStringValue("NickName");
    else:
        urlParam.Result = False;
    return urlParam;

def takeAction(urlParam, parent):
    actionResult = ActionResult();
    userId = parent.Current.User.PersonalId;
    contextUser = parent.Current.User;
    nickName = urlParam.nickName.lstrip().rstrip();

    MaxLength = ConfigEnvSet.GetInt("User.MaxLength");
    itemID = ConfigEnvSet.GetInt("UserName.ItemID");
    length = Text.Encoding.Default.GetByteCount(nickName);
    if length <= 0 or length > MaxLength:
        parent.ErrorCode = LanguageManager.GetLang().ErrorCode;
        parent.ErrorInfo = LanguageManager.GetLang().St1005_KingNameTooLong.format(MaxLength);
        actionResult.Result = False;
        return actionResult;
    # 屏蔽敏感词
    if UserHelper.GetKeyWordSubstitution(nickName):
        parent.ErrorCode = LanguageManager.GetLang().ErrorCode;
        parent.ErrorInfo = LanguageManager.GetLang().St1005_RegistNameKeyWord;
        actionResult.Result = False;
        return actionResult;

    # 是否已经存在该昵称
    if GameUser.IsNickName(nickName):
        parent.ErrorCode = 1;
        parent.ErrorInfo = LanguageManager.GetLang().St1005_Rename;
        actionResult.Result = False;
        return actionResult;
    package = UserItemPackage.Get(userId);
    userItem = package.ItemPackage.Find(match=lambda m:not m.IsRemove and m.ItemID == itemID);
    if userItem == None or userItem.Num <= 0:
        parent.ErrorCode = LanguageManager.GetLang().ErrorCode;
        actionResult.Result = False;
        return actionResult;
    contextUser.NickName = nickName;
        
    ranking = RankingFactory.Get[UserRank](CombatRanking.RankingKey);
    if ranking.TryGetRankNo(match=lambda m:m.UserID == userId):
        rankInfo = ranking.Find(match=lambda s:s.UserID == userId);
        if rankInfo:
            rankInfo.NickName = nickName;
    UserItemHelper.UseUserItem(userId, itemID, 1);

    return actionResult;

def buildPacket(writer, urlParam, actionResult):
    return True 

File: Sample/Koudai/Server/src/ZyGames.Tianjiexing.BLL/Action/Action1023.cs
/****************************************************************************
Copyright (c) 2013-2015 scutgame.com

http://www.scutgame.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System;

using System.Data;
using ZyGames.Framework.Common;
using ZyGames.Framework.Game.Com;
using ZyGames.Framework.Game.Com.Rank;
using ZyGames.Framework.Game.Service;
using ZyGames.Tianjiexing.Component;
using ZyGames.Tianjiexing.Model;
using ZyGames.Tianjiexing.BLL.Base;
using ZyGames.Tianjiexing.Model.Config;
using ZyGames.Tianjiexing.Lang;
using System.Text;

namespace ZyGames.Tianjiexing.BLL.Action
{

    /// <summary>
    /// 1023_玩家改名
    /// </summary>
    public class Action1023 : BaseAction
    {
        private int MaxLength = 0;
        private string _nickName = string.Empty;
        private int itemID = 0;

        public Action1023(ZyGames.Framework.Game.Contract.HttpGet httpGet)
            : base(ActionIDDefine.Cst_Action1023, httpGet)
        {

        }

        public override void BuildPacket()
        {

        }

        public override bool GetUrlElement()
        {
            if (httpGet.GetString("NickName", ref _nickName))
            {
                _nickName = _nickName.Trim();
                return true;
            }
            return false;
        }

        public override bool TakeAction()
        {
            try
            {
                MaxLength = ConfigEnvSet.GetInt("User.MaxLength");
                itemID = ConfigEnvSet.GetInt("UserName.ItemID");
                int length = System.Text.Encoding.Default.GetByteCount(_nickName);
                if (length <= 0 || length > MaxLength)
                {
                    ErrorCode = LanguageManager.GetLang().ErrorCode;
                    ErrorInfo = string.Format(LanguageManager.GetLang().St1005_KingNameTooLong, MaxLength);
                    return false;
                }
                if (GameUser.IsNickName(_nickName))
                {
                    ErrorCode = 1;
                    ErrorInfo = LanguageManager.GetLang().St1005_Rename;
                    return false;
                }
                //UserItemLog userItem = UserItemLog.FindKey(itemID);
                var package = UserItemPackage.Get(Uid);
                UserItemInfo userItem = package.ItemPackage.Find(m => !m.IsRemove && m.ItemID == itemID);
                if (userItem == null || userItem.Num <= 0)
                {
                    ErrorCode = LanguageManager.GetLang().ErrorCode;
                    return false;
                }
                ContextUser.NickName = _nickName;

                int rankID = 0;
                UserRank rankInfo;
                Ranking<UserRank> ranking = RankingFactory.Get<UserRank>(CombatRanking.RankingKey);
                if (ranking.TryGetRankNo(m => m.UserID == ContextUser.UserID, out rankID))
                {
                    rankInfo = ranking.Find(s => s.UserID == ContextUser.UserID);
                    if (rankInfo != null)
                    {
                        rankInfo.NickName = _nickName;
                    }
                }
                //ContextUser.Update();
                //UserGeneral usergen = UserGeneral.GetMainGeneral(ContextUser.UserID);
                //if (usergen != null)
                //{
                //    usergen.GeneralName = _nickName;
                //    //usergen.Update();
                //}
                UserItemHelper.UseUserItem(ContextUser.UserID, itemID, 1);

                return true;
            }
            catch (Exception ex)
            {
                SaveLog(ex);
                return false;
            }
        }
    }
}